File: src/test/kotlin/playground/ttl/deepcopy/TtlDeepCopyDemo.java
package playground.ttl.deepcopy;

import com.alibaba.ttl.TransmittableThreadLocal;

import java.util.HashSet;
import java.util.Set;

public class TtlDeepCopyDemo {
    public static void main(String[] args) throws Exception {
        context.get().add(new Person("cat", 18));

        Thread thread = new Thread(() -> System.out.println(context.get()));
        thread.start();
        thread.join();
    }

    private static final TransmittableThreadLocal<Set<Person>> context = new TransmittableThreadLocal<Set<Person>>() {
        @Override
        public Set<Person> copy(Set<Person> parentValue) {
            Set<Person> deep = new HashSet<>();
            for (Person person : parentValue) {
                deep.add(new Person(person.name + "+1", person.age + 1));
            }
            return deep;
        }

        @Override
        protected Set<Person> childValue(Set<Person> parentValue) {
            return copy(parentValue);
        }

        @Override
        protected Set<Person> initialValue() {
            return new HashSet<>();
        }
    };

    private static class Person {
        private final String name;
        private final int age;

        public Person(String name, int age) {
            this.name = name;
            this.age = age;
        }

        @Override
        public String toString() {
            return "Person{name='" + name + '\'' + ", age=" + age + '}';
        }
    }
}

File: src/test/kotlin/playground/ttl/deepcopy/TtlDeepCopyDemo.kt
package playground.ttl.deepcopy

import com.alibaba.ttl.TransmittableThreadLocal
import kotlin.concurrent.thread

fun main() {
    context.get().add(Person("cat", 18))

    thread {
        println(context.get())
    }.join()
}

private val context: TransmittableThreadLocal<MutableSet<Person>> =
    object : TransmittableThreadLocal<MutableSet<Person>>() {
        override fun copy(parentValue: MutableSet<Person>): MutableSet<Person> = copyByJavaFor(parentValue)

        override fun childValue(parentValue: MutableSet<Person>): MutableSet<Person> = copyByJavaFor(parentValue)

        override fun initialValue(): MutableSet<Person> = hashSetOf()
    }

private data class Person(val name: String, val age: Int)


////////////////////////////////////////////////////////////////////////////////


/**
 * use java old school for statement.
 *
 * high performance for both big and small set
 */
private fun copyByJavaFor(parentValue: MutableSet<Person>): MutableSet<Person> = hashSetOf<Person>().also {
    for (person in parentValue) {
        it.add(person.copy(name = "${person.name}+1", age = person.age + 1))
    }
}

/**
 * use simple and direct kotlin style.
 *
 * - poor performance for big set
 * - high performance for small set
 */
private fun copyByKotlinStyle(parentValue: MutableSet<Person>): MutableSet<Person> =
    parentValue.map { person ->
        person.copy(name = "${person.name}+1", age = person.age + 1)
    }.toHashSet()

/**
 * use Kotlin [Sequence].
 *
 * - high performance for big set
 * - poor performance for small set
 */
private fun copyByKotlinSequence(parentValue: MutableSet<Person>): MutableSet<Person> =
    parentValue.asSequence().map { person ->
        person.copy(name = "${person.name}+1", age = person.age + 1)
    }.toCollection(hashSetOf())
